README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Prometheus components for Dart

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module contains components for working with meters in the Prometheus service. The PrometheusCounters and PrometheusMetricsService components allow you to work both in client mode through PushGateway, and as a service.

The module contains the following packages:
- **Build** - the default factories for constructing components.
- **Count** - components of counters (metrics) with sending data to Prometheus via PushGateway
- **Services** - components of the service for reading counters (metrics) by the Prometheus service

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pub.dev/documentation/pip_services3_prometheus/latest/pip_services3_prometheus/pip_services3_prometheus-library.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)


## Use

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  pip_services3_prometheus: version
```

Now you can install package from the command line:
```bash
pub get
```

## Develop

For development you shall install the following prerequisites:
* Dart SDK 2
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pub get
```

Run automated tests:
```bash
pub run test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by 
- **Sergey Seroukhov**
- **Levichev Dmitry**

The documentation is written by:
- **Levichev Dmitry**
- **Mark Makarychev**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Dart Changelog

## 1.0.0

- Initial version, created by Sergey Seroukhov and Dmitry Levichev

## 1.1.0

- Migrated to null safety and **shelf** framework

test/fixtures/CountersFixture.dart
import 'dart:async';

import 'package:test/test.dart';

import 'package:pip_services3_components/pip_services3_components.dart';

class CountersFixture {
  final CachedCounters _counters;

  CountersFixture(CachedCounters counters) : _counters = counters;

  Future testSimpleCounters() async {
    _counters.last('Test.LastValue', 123);
    _counters.last('Test.LastValue', 123456);

    var counter = _counters.get('Test.LastValue', CounterType.LastValue);
    expect(counter, isNotNull);
    expect(counter.last, isNotNull);
    expect(counter.last, 123456); //, 3

    _counters.incrementOne('Test.Increment');
    _counters.increment('Test.Increment', 3);

    counter = _counters.get('Test.Increment', CounterType.Increment);
    expect(counter, isNotNull);
    expect(counter.count, 4);

    _counters.timestampNow('Test.Timestamp');
    _counters.timestampNow('Test.Timestamp');

    counter = _counters.get('Test.Timestamp', CounterType.Timestamp);
    expect(counter, isNotNull);
    expect(counter.time, isNotNull);

    _counters.stats('Test.Statistics', 1);
    _counters.stats('Test.Statistics', 2);
    _counters.stats('Test.Statistics', 3);

    counter = _counters.get('Test.Statistics', CounterType.Statistics);
    expect(counter, isNotNull);
    expect(counter.average, 2); // , 3

    _counters.dump();

    await Future.delayed(Duration(milliseconds: 1000));
  }

  Future testMeasureElapsedTime() async {
    var timer = _counters.beginTiming('Test.Elapsed');

    Future.delayed(Duration(milliseconds: 100), () async {
      timer.endTiming();

      var counter = _counters.get('Test.Elapsed', CounterType.Interval);
      expect(counter.last! > 50, isTrue);
      expect(counter.last! < 5000, isTrue);

      _counters.dump();

      await Future.delayed(Duration(milliseconds: 1000));
    });
  }
}


test/count/PrometheusCounters.test.dart
// import { ConfigParams } from 'pip-services3-commons-node';
// import { References } from 'pip-services3-commons-node';

// import { PrometheusCounters } from '../../src/count/PrometheusCounters';
// import { CountersFixture } from '../fixtures/CountersFixture';

// suite('PrometheusCounters', ()=> {
//     let _counters: PrometheusCounters;
//     let _fixture: CountersFixture;

//     setup((done) => {
//         let host = process.env['PUSHGATEWAY_SERVICE_HOST'] || 'localhost';
//         let port = process.env['PUSHGATEWAY_SERVICE_PORT'] || 9091;

//         _counters = new PrometheusCounters();
//         _fixture = new CountersFixture(_counters);

//         let config = ConfigParams.fromTuples(
//             'source', 'test',
//             'connection.host', host,
//             'connection.port', port
//         );
//         _counters.configure(config);

//         _counters.open(null, (err) => {
//              done(err);
//         });
//     });

//     teardown((done) => {
//         _counters.close(null, done);
//     });

//     test('Simple Counters', (done) => {
//         _fixture.testSimpleCounters(done);
//     });

//     test('Measure Elapsed Time', (done) => {
//         _fixture.testMeasureElapsedTime(done);
//     });

// });

test/count/PrometheusCounters_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';

import 'package:pip_services3_prometheus/src/count/PrometheusCounters.dart';
import '../fixtures/CountersFixture.dart';

void main() {
  group('PrometheusCounters', () {
    late PrometheusCounters _counters;
    late CountersFixture _fixture;

    setUp(() async {
      var host =
          Platform.environment['PUSHGATEWAY_SERVICE_HOST'] ?? 'localhost';
      var port = Platform.environment['PUSHGATEWAY_SERVICE_PORT'] ?? 9091;

      _counters = PrometheusCounters();
      _fixture = CountersFixture(_counters);

      var config = ConfigParams.fromTuples(
          ['source', 'test', 'connection.host', host, 'connection.port', port]);
      _counters.configure(config);

      await _counters.open(null);
    });

    tearDown(() async {
      await _counters.close(null);
    });

    test('Simple Counters', () async {
      await _fixture.testSimpleCounters();
    });

    test('Measure Elapsed Time', () async {
      await _fixture.testMeasureElapsedTime();
    });
  });
}


example/main.dart
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

import 'package:pip_services3_prometheus/pip_services3_prometheus.dart';

void main() async {
  var restConfig = ConfigParams.fromTuples([
    'connection.protocol',
    'http',
    'connection.host',
    'localhost',
    'connection.port',
    3000
  ]);

  PrometheusMetricsService service;
  PrometheusCounters counters;
 
  service = PrometheusMetricsService();
  counters = PrometheusCounters();

  service.configure(restConfig);

  var contextInfo = ContextInfo();
  contextInfo.name = 'Test';
  contextInfo.description = 'This is a test container';

  var references = References.fromTuples([
    Descriptor('pip-services', 'context-info', 'default', 'default', '1.0'),
    contextInfo,
    Descriptor('pip-services', 'counters', 'prometheus', 'default', '1.0'),
    counters,
    Descriptor(
        'pip-services', 'metrics-service', 'prometheus', 'default', '1.0'),
    service
  ]);
  counters.setReferences(references);
  service.setReferences(references);

  await counters.open(null);
  await service.open(null);

  counters.incrementOne('test.counter1');
  counters.stats('test.counter2', 2);
  counters.last('test.counter3', 3);
  counters.timestampNow('test.counter4');

  // all metrics accessable on http://localhost:3000/metrics
  // configure the Prometheus service to poll this host
  // ...

  await service.close(null);
  await counters.close(null);
}


lib/src/count/PrometheusCounters.dart
import 'dart:async';
import 'dart:io';

import 'package:http/http.dart' as http;

import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';
import 'package:pip_services3_rpc/pip_services3_rpc.dart';
import './PrometheusCounterConverter.dart';

/// Performance counters that send their metrics to Prometheus service.
///
/// The component is normally used in passive mode conjunction with [PrometheusMetricsService].
/// Alternatively when connection parameters are set it can push metrics to Prometheus PushGateway.
///
/// ### Configuration parameters ###
///
/// - [connection(s)]:
///   - [discovery_key]:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///   - [protocol]:              connection protocol: http or https
///   - [host]:                  host name or IP address
///   - [port]:                  port number
///   - [uri]:                   resource URI or connection string with all parameters in it
/// - [options]:
///   - [retries]:               number of retries (default: 3)
///   - [connect_timeout]:       connection timeout in milliseconds (default: 10 sec)
///   - [timeout]:               invocation timeout in milliseconds (default: 10 sec)
///
/// ### References ###
///
/// - *:logger:*:*:1.0         (optional) [ILogger](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ILogger-class.html) components to pass log messages
/// - *:counters:*:*:1.0         (optional) [ICounters](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICounters-class.html) components to pass collected measurements
/// - *:discovery:*:*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
///
/// See [RestService](https://pub.dev/documentation/pip_services3_rpc/latest/pip_services3_rpc/RestService-class.html)
/// See [CommandableHttpService](https://pub.dev/documentation/pip_services3_rpc/latest/pip_services3_rpc/CommandableHttpService-class.html)
///
/// ### Example ###
///
///     var counters =  PrometheusCounters();
///     counters.configure(ConfigParams.fromTuples([
///         'connection.protocol', 'http',
///         'connection.host', 'localhost',
///         'connection.port', 8080
///     ]));
///
///     await counters.open('123')
///         ...
///
///     counters.increment('mycomponent.mymethod.calls');
///     var timing = counters.beginTiming('mycomponent.mymethod.exec_time');
///     try {
///         ...
///     } finally {
///         timing.endTiming();
///     }
///
///     counters.dump();

class PrometheusCounters extends CachedCounters
    implements IReferenceable, IOpenable {
  final _logger = CompositeLogger();
  final _connectionResolver = HttpConnectionResolver();
  bool _opened = false;
  String? _source;
  String? _instance;
  bool _pushEnabled = true;
  http.Client? _client;
  String? _requestRoute;
  String? _uri;

  /// Creates a new instance of the performance counters.
  PrometheusCounters() : super();

  /// Configures component by passing configuration parameters.
  ///
  /// - [config]    configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    super.configure(config);

    _connectionResolver.configure(config);
    _source = config.getAsNullableString('source') ?? _source;
    _instance = config.getAsNullableString('instance') ?? _instance;
    _pushEnabled = config.getAsBooleanWithDefault('push_enabled', true);
  }

  /// Sets references to dependent components.
  ///
  /// - [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    _logger.setReferences(references);
    _connectionResolver.setReferences(references);

    var contextInfo = references.getOneOptional<ContextInfo>(
        Descriptor('pip-services', 'context-info', 'default', '*', '1.0'));
    if (contextInfo != null && _source == null) {
      _source = contextInfo.name;
    }
    if (contextInfo != null && _instance == null) {
      _instance = contextInfo.contextId;
    }
  }

  /// Checks if the component is opened.
  ///
  /// Returns true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _opened;
  }

  /// Opens the component.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives null no errors occured.
  /// Throws error
  @override
  Future open(String? correlationId) async {
    if (_opened) {
      return null;
    }
    if (!_pushEnabled) {
      return null;
    }

    ConfigParams? connection;

    try {
      connection = await _connectionResolver.resolve(correlationId);
      if (connection == null) {
        throw Exception('Empty config.');
      }
    } catch (err) {
      _client = null;
      _logger.warn(
          correlationId,
          'Connection to Prometheus server is not configured: ' +
              err.toString());
      return null;
    }
    var job = _source ?? 'unknown';
    var instance = _instance ?? Platform.localHostname;
    _requestRoute = '/metrics/job/' + job + '/instance/' + instance;
    _uri = connection.getAsString('uri');
    _client = http.Client();
    _opened = true;
  }

  /// Closes component and frees used resources.
  ///
  /// - [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			      Future that receives null no errors occured.
  /// Throws error
  @override
  Future close(String? correlationId) async {
    _opened = false;
    if (_client != null) {
      _client!.close();
    }
    _client = null;
    _requestRoute = null;
  }

  /// Saves the current counters measurements.
  ///
  /// - [counters]      current counters measurements to be saves.
  @override
  Future save(List<Counter> counters) async {
    if (_client == null || !_pushEnabled) return;
    var body = PrometheusCounterConverter.toString2(counters, null, null);

    var url = _uri! + _requestRoute!;
    try {
      var response = await _client!
          .put(Uri.parse(url), headers: {'Accept': 'text/html'}, body: body);
      if (response.statusCode >= 400) {
        _logger.error('prometheus-counters', ApplicationException(),
            'Failed to push metrics to prometheus');
      }
    } catch (ex) {
      _logger.error('prometheus-counters', ApplicationException().wrap(ex),
          'Failed to push metrics to prometheus');
    }
  }
}


lib/src/count/PrometheusCounterConverter.dart
import 'package:pip_services3_components/pip_services3_components.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';

/// Helper class that converts performance counter values into
/// a response from Prometheus metrics service.

class PrometheusCounterConverter {
  /// Converts the given counters to a string that is returned by Prometheus metrics service.
  ///
  /// - [counters]  a list of counters to convert.
  /// - [source]    a source (context) name.
  /// - [instance]  a unique instance name (usually a host name).
  static String toString2(
      List<Counter>? counters, String? source, String? instance) {
    if (counters == null || counters.isEmpty) return '';

    var builder = '';

    for (var counter in counters) {
      var counterName = _parseCounterName(counter);
      var labels = _generateCounterLabel(counter, source, instance);

      switch (counter.type) {
        case CounterType.Increment:
          builder += '# TYPE ' + counterName + ' gauge\n';
          builder += counterName +
              labels +
              ' ' +
              StringConverter.toString2(counter.count) +
              '\n';
          break;
        case CounterType.Interval:
          builder += '# TYPE ' + counterName + '_max gauge\n';
          builder += counterName +
              '_max' +
              labels +
              ' ' +
              StringConverter.toString2(counter.max) +
              '\n';
          builder += '# TYPE ' + counterName + '_min gauge\n';
          builder += counterName +
              '_min' +
              labels +
              ' ' +
              StringConverter.toString2(counter.min) +
              '\n';
          builder += '# TYPE ' + counterName + '_average gauge\n';
          builder += counterName +
              '_average' +
              labels +
              ' ' +
              StringConverter.toString2(counter.average) +
              '\n';
          builder += '# TYPE ' + counterName + '_count gauge\n';
          builder += counterName +
              '_count' +
              labels +
              ' ' +
              StringConverter.toString2(counter.count) +
              '\n';
          break;
        case CounterType.LastValue:
          builder += '# TYPE ' + counterName + ' gauge\n';
          builder += counterName +
              labels +
              ' ' +
              StringConverter.toString2(counter.last) +
              '\n';
          break;
        case CounterType.Statistics:
          builder += '# TYPE ' + counterName + '_max gauge\n';
          builder += counterName +
              '_max' +
              labels +
              ' ' +
              StringConverter.toString2(counter.max) +
              '\n';
          builder += '# TYPE ' + counterName + '_min gauge\n';
          builder += counterName +
              '_min' +
              labels +
              ' ' +
              StringConverter.toString2(counter.min) +
              '\n';
          builder += '# TYPE ' + counterName + '_average gauge\n';
          builder += counterName +
              '_average' +
              labels +
              ' ' +
              StringConverter.toString2(counter.average) +
              '\n';
          builder += '# TYPE ' + counterName + '_count gauge\n';
          builder += counterName +
              '_count' +
              labels +
              ' ' +
              StringConverter.toString2(counter.count) +
              '\n';
          break;
        case CounterType
            .Timestamp: // Prometheus doesn't support non-numeric metrics
          builder += '# TYPE ' + counterName + ' gauge\n'; //' untyped\n';
          builder += counterName +
              labels +
              ' ' +
              StringConverter.toString2(counter.time?.millisecondsSinceEpoch) +
              '\n';
          break;
      }
    }

    return builder;
  }

  static String _generateCounterLabel(
      Counter counter, String? source, String? instance) {
    var labels = <String, dynamic>{};

    if (source != null && source != '') labels['source'] = source;
    if (instance != null && instance != '') labels['instance'] = instance;

    var nameParts = counter.name.split('.');

    // If there are other predictable names from which we can parse labels, we can add them below
    if ((nameParts.length >= 3 && nameParts[2] == 'exec_count') ||
        (nameParts.length >= 3 && nameParts[2] == 'exec_time') ||
        (nameParts.length >= 3 && nameParts[2] == 'exec_errors')) {
      labels['service'] = nameParts[0];
      labels['command'] = nameParts[1];
    }

    if ((nameParts.length >= 4 && nameParts[3] == 'call_count') ||
        (nameParts.length >= 4 && nameParts[3] == 'call_time') ||
        (nameParts.length >= 4 && nameParts[3] == 'call_errors')) {
      labels['service'] = nameParts[1];
      labels['command'] = nameParts[2];
      labels['target'] = nameParts[0];
    }

    if ((nameParts.length >= 3 && nameParts[2] == 'sent_messages') ||
        (nameParts.length >= 3 && nameParts[2] == 'received_messages') ||
        (nameParts.length >= 3 && nameParts[2] == 'dead_messages')) {
      labels['queue'] = nameParts[1];
    }

    if (labels.isEmpty) return '';

    var builder = '{';
    for (var key in labels.keys) {
      if (builder.length > 1) builder += ',';
      builder += key + '="' + labels[key] + '"';
    }
    builder += '}';

    return builder;
  }

  static String _parseCounterName(Counter? counter) {
    if (counter?.name == null && counter?.name == '' && counter == null) {
      return '';
    }

    var nameParts = counter!.name.split('.');

    // If there are other predictable names from which we can parse labels, we can add them below
    // Rest Service Labels
    if (nameParts.length >= 3 && nameParts[2] == 'exec_count') {
      return nameParts[2];
    }
    if (nameParts.length >= 3 && nameParts[2] == 'exec_time') {
      return nameParts[2];
    }
    if (nameParts.length >= 3 && nameParts[2] == 'exec_errors') {
      return nameParts[2];
    }

    // Rest & Direct Client Labels
    if (nameParts.length >= 4 && nameParts[3] == 'call_count') {
      return nameParts[3];
    }
    if (nameParts.length >= 4 && nameParts[3] == 'call_time') {
      return nameParts[3];
    }
    if (nameParts.length >= 4 && nameParts[3] == 'call_errors') {
      return nameParts[3];
    }

    // Queue Labels
    if ((nameParts.length >= 3 && nameParts[2] == 'sent_messages') ||
        (nameParts.length >= 3 && nameParts[2] == 'received_messages') ||
        (nameParts.length >= 3 && nameParts[2] == 'dead_messages')) {
      var name = '${nameParts[0]}.${nameParts[2]}';
      return name.toLowerCase().replaceAll('.', '_').replaceAll('/', '_');
    }

    // TODO: are there other assumptions we can make?
    // Or just return as a single, valid name
    return counter.name.toLowerCase().replaceAll('.', '_').replaceAll('/', '_');
  }

  static Map<String, dynamic> _parseCounterLabels(
      Counter counter, String? source, String? instance) {
    var labels = <String, dynamic>{};

    if (source != null && source != '') labels['source'] = source;
    if (instance != null && instance != '') labels['instance'] = instance;

    var nameParts = counter.name.split('.');

    // If there are other predictable names from which we can parse labels, we can add them below
    if (nameParts.length >= 3 && nameParts[2] == 'exec_time') {
      labels['service'] = nameParts[0];
      labels['command'] = nameParts[1];
    }

    return labels;
  }
}


lib/src/services/PrometheusMetricsService.dart
import 'dart:async';

import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_rpc/pip_services3_rpc.dart';
import 'package:pip_services3_components/pip_services3_components.dart';
import 'package:shelf/shelf.dart';

import '../count/PrometheusCounters.dart';
import '../count/PrometheusCounterConverter.dart';

/// Service that exposes the '/metrics' and '/metricsandreset' routes for Prometheus to scap performance metrics.
///
/// ### Configuration parameters ###
///
/// - [dependencies]:
///   - [endpoint]:              override for HTTP Endpoint dependency
///   - [prometheus-counters]:   override for PrometheusCounters dependency
/// - [connection(s)]:
///   - [discovery_key]:         (optional) a key to retrieve the connection from IDiscovery
///   - [protocol]:              connection protocol: http or https
///   - [host]:                  host name or IP address
///   - [port]:                  port number
///   - [uri]:                   resource URI or connection string with all parameters in it
///
/// ### References ###
///
/// - *:logger:*:*:1.0         (optional) [ILogger](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ILogger-class.html) components to pass log messages
/// - *:counters:*:*:1.0         (optional) [ICounters](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ICounters-class.html) components to pass collected measurements
/// - *:discovery:*:*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
/// - *:endpoint:http:*:1.0          (optional) [HttpEndpoint](https://pub.dev/documentation/pip_services3_rpc/latest/pip_services3_rpc/HttpEndpoint-class.html) reference to expose REST operation
/// - *:counters:prometheus:*:1.0    [PrometheusCounters] reference to retrieve collected metrics
///
/// See [RestService](https://pub.dev/documentation/pip_services3_rpc/latest/pip_services3_rpc/RestService-class.html)
/// See [RestClient](https://pub.dev/documentation/pip_services3_rpc/latest/pip_services3_rpc/RestClient-class.html)
///
/// ### Example ###
///
///     var service = PrometheusMetricsService();
///     service.configure(ConfigParams.fromTuples([
///         'connection.protocol', 'http',
///         'connection.host', 'localhost',
///         'connection.port', 8080
///     ]));
///
///     await service.open('123')
///     print('The Prometheus metrics service is accessible at http://+:8080/metrics');

class PrometheusMetricsService extends RestService {
  CachedCounters? _cachedCounters;
  String? _source;
  String? _instance;

  /// Creates a new instance of this service.
  PrometheusMetricsService() : super() {
    dependencyResolver.put('cached-counters',
        Descriptor('pip-services', 'counters', 'cached', '*', '1.0'));
    dependencyResolver.put('prometheus-counters',
        Descriptor('pip-services', 'counters', 'prometheus', '*', '1.0'));
  }

  /// Sets references to dependent components.
  ///
  /// - [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    super.setReferences(references);

    _cachedCounters = dependencyResolver
        .getOneOptional<PrometheusCounters>('prometheus-counters');
    _cachedCounters ??=
        dependencyResolver.getOneOptional<CachedCounters>('cached-counters');

    var contextInfo = references.getOneOptional<ContextInfo>(
        Descriptor('pip-services', 'context-info', 'default', '*', '1.0'));

    if (contextInfo != null && (_source == '' || _source == null)) {
      _source = contextInfo.name;
    }
    if (contextInfo != null && (_instance == '' || _instance == null)) {
      _instance = contextInfo.contextId;
    }
  }

  /// Registers all service routes in HTTP endpoint.
  @override
  void register() {
    registerRoute('get', 'metrics', null, (Request req) async {
      return await _metrics(req);
    });
    registerRoute('get', 'metricsandreset', null, (Request req) async {
      return await _metricsAndReset(req);
    });
  }

  /// Handles metrics requests
  ///
  /// - [req]   an HTTP request
  /// - [res]   an HTTP response
  FutureOr<Response> _metrics(Request req) {
    var counters = _cachedCounters != null ? _cachedCounters!.getAll() : null;
    var body =
        PrometheusCounterConverter.toString2(counters, _source, _instance);

    return Response(200, headers: {'content-type': 'text/plain'}, body: body);
  }

  /// Handles metricsandreset requests.
  /// The counters will be returned and then zeroed out.
  /// - [req]   an HTTP request
  /// - [res]   an HTTP response
  FutureOr<Response> _metricsAndReset(Request req) {
    var counters = _cachedCounters != null ? _cachedCounters!.getAll() : null;
    var body =
        PrometheusCounterConverter.toString2(counters, _source, _instance);

    if (_cachedCounters != null) {
      _cachedCounters!.clearAll();
    }

    return Response(200, headers: {'content-type': 'text/plain'}, body: body);
  }
}
